File: Sources/StitchPilotCore/Licensing/Entitlement.swift
import CryptoKit
import Foundation

/// A signed statement from License Admin: "customer N, on device D, may
/// use PiperStitch until `expiresAt`." The subscription-world replacement
/// for a license key — short-lived, refreshed silently, and verified here
/// against the public key in `LicenseConfig` so the app never has to trust
/// the network or a value on disk.
///
/// Wire format, kept byte-for-byte in sync with license-admin's
/// app/entitlement.py:
///
///     PSE1.<base64url(payload JSON, no padding)>.<base64url(64-byte Ed25519 signature)>
///
/// The signature covers the exact payload bytes; verification happens
/// BEFORE the JSON is parsed, and nothing in the payload is trusted until
/// it has.
public struct Entitlement: Equatable, Sendable {
    public let customerID: Int
    public let deviceID: String
    public let email: String
    /// 'active' | 'trialing' | 'past_due' | 'comp' — for display only.
    public let status: String
    public let expiresAt: Date
    public let issuedAt: Date
    /// The billing period's real end (nil for a comp with no end set).
    public let periodEnd: Date?
    public let cancelAtPeriodEnd: Bool
    /// The token exactly as received, for storage and re-verification.
    public let token: String

    public func isValid(at now: Date = Date()) -> Bool { now < expiresAt }
}

public enum EntitlementError: Error, Equatable, LocalizedError {
    case malformed
    case badSignature
    case unsupportedVersion
    case corruptPayload
    case wrongDevice
    case badPublicKey

    public var errorDescription: String? {
        switch self {
        case .malformed: return "That isn't a PiperStitch entitlement."
        case .badSignature: return "This entitlement isn't valid."
        case .unsupportedVersion: return "This entitlement was issued in a newer format than this version of PiperStitch understands. Please update the app."
        case .corruptPayload: return "This entitlement is corrupted."
        case .wrongDevice: return "This entitlement was issued to a different Mac."
        case .badPublicKey: return "The app's licensing key is misconfigured."
        }
    }
}

public enum EntitlementVerifier {
    static let prefix = "PSE1"
    static let version = 1

    private struct Payload: Decodable {
        let v: Int
        let cid: Int
        let dev: String
        let email: String
        let status: String
        let exp: String
        let iat: String
        let period_end: String?
        let cancel_at_period_end: Bool?
    }

    /// Verifies the signature and returns the entitlement — or throws.
    /// `expectedDeviceID` binds the token to this Mac: a token copied from
    /// another machine fails even though its signature is genuine.
    public static func verify(_ token: String, publicKeyBase64: String, expectedDeviceID: String? = nil) throws -> Entitlement {
        let trimmed = token.trimmingCharacters(in: .whitespacesAndNewlines)
        let parts = trimmed.split(separator: ".", omittingEmptySubsequences: false).map(String.init)
        guard parts.count == 3, parts[0] == prefix,
              let payloadBytes = base64URLDecode(parts[1]),
              let signature = base64URLDecode(parts[2]), signature.count == 64 else {
            throw EntitlementError.malformed
        }
        guard let keyBytes = Data(base64Encoded: publicKeyBase64),
              let publicKey = try? Curve25519.Signing.PublicKey(rawRepresentation: keyBytes) else {
            throw EntitlementError.badPublicKey
        }
        guard publicKey.isValidSignature(signature, for: payloadBytes) else {
            throw EntitlementError.badSignature
        }

        let payload: Payload
        do {
            payload = try JSONDecoder().decode(Payload.self, from: payloadBytes)
        } catch {
            throw EntitlementError.corruptPayload
        }
        guard payload.v == version else { throw EntitlementError.unsupportedVersion }
        guard let exp = parseISO8601(payload.exp), let iat = parseISO8601(payload.iat) else {
            throw EntitlementError.corruptPayload
        }
        if let expected = expectedDeviceID, expected != payload.dev {
            throw EntitlementError.wrongDevice
        }
        return Entitlement(
            customerID: payload.cid,
            deviceID: payload.dev,
            email: payload.email,
            status: payload.status,
            expiresAt: exp,
            issuedAt: iat,
            periodEnd: payload.period_end.flatMap(parseISO8601),
            cancelAtPeriodEnd: payload.cancel_at_period_end ?? false,
            token: trimmed
        )
    }

    static func base64URLDecode(_ s: String) -> Data? {
        var b64 = s.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        while b64.count % 4 != 0 { b64 += "=" }
        return Data(base64Encoded: b64)
    }

    static func parseISO8601(_ s: String) -> Date? {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        if let d = f.date(from: s) { return d }
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f.date(from: s)
    }
}

File: license-admin/app/entitlement.py
"""Entitlement tokens — the subscription-world replacement for Amerus's
license key.

Amerus mints one permanent Ed25519-signed key per purchase and the app
verifies it forever. PiperStitch is a monthly subscription, so what the
app holds instead is a short-lived *entitlement*: a signed statement
"customer N, on device D, may use PiperStitch until date X". The app
refreshes it silently whenever it has a network connection (see
Sources/StitchPilotCore/Licensing/ in the main repo); if it can't, the
one it already has keeps working until X. X is bounded by
config.ENTITLEMENT_MAX_DAYS so a cancelled subscription can't stay
unlocked indefinitely by going offline — see subscriptions.validity_for.

Wire format (must stay byte-for-byte in sync with the Swift verifier,
EntitlementVerifier.swift):

    PSE1.<base64url(payload JSON, no padding)>.<base64url(64-byte Ed25519 signature)>

The signature is over the exact payload bytes as encoded — the app
verifies first, then parses the JSON, and never trusts a field it
couldn't verify. The payload is deliberately plain JSON rather than the
packed binary Amerus uses: nobody ever types an entitlement by hand, so
its length doesn't matter, and JSON is trivially decodable on the Swift
side with no custom struct-packing to keep in sync.

Payload fields:
    v        1 — format version; anything else is rejected
    cid      customer id in this service's database
    dev      the device_id the app presented at sign-in (the app refuses
             a token minted for a different Mac)
    email    for display in the app's account panel only
    status   'active' | 'trialing' | 'past_due' | 'comp' — display only
    exp      ISO-8601 UTC — the app locks after this
    iat      ISO-8601 UTC — issued at
    period_end   ISO-8601 UTC or null — the billing period's real end,
             so the app can say "renews on ..." / "ends on ..."
    cancel_at_period_end   bool — so the app can say "ends" not "renews"

Like the Amerus License Admin's keygen.py, this module is handed the
PRIVATE key (config.PIPERSTITCH_LICENSE_PRIVATE_KEY): never log it, never
put it in an error message, never send it anywhere but the signing call.
"""

from __future__ import annotations

import base64
import binascii
import json
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Optional

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey

TOKEN_PREFIX = "PSE1"
_VERSION = 1

# The public half of the keypair in ~/Documents/PiperStitch-Licensing/ —
# the same value baked into the shipped app's EntitlementVerifier.swift.
# Safe to keep here: it can confirm a token is genuine but can't mint one.
# Used only to self-check freshly-signed tokens before they leave.
_PUBLIC_KEY_B64 = "OLvbWOM6exl9IL11JHluZkzrTVsQQZs7kB3eHpJJJDY="


class EntitlementError(Exception):
    """Signing failed (bad/missing private key) or a token failed to
    verify — always a plain-language reason, never a raw crypto exception."""


@dataclass(frozen=True)
class Entitlement:
    customer_id: int
    device_id: str
    email: str
    status: str
    expires_at: datetime
    issued_at: datetime
    period_end: Optional[datetime]
    cancel_at_period_end: bool
    canonical: str


def _b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def _unb64url(text: str) -> bytes:
    return base64.urlsafe_b64decode(text + "=" * (-len(text) % 4))


def _iso(dt: datetime) -> str:
    return dt.astimezone(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def _parse_iso(value: str) -> datetime:
    return datetime.fromisoformat(value.replace("Z", "+00:00")).astimezone(timezone.utc)


def _public_key() -> Ed25519PublicKey:
    return Ed25519PublicKey.from_public_bytes(base64.b64decode(_PUBLIC_KEY_B64))


def sign(
    private_key_b64: str,
    *,
    customer_id: int,
    device_id: str,
    email: str,
    status: str,
    expires_at: datetime,
    period_end: Optional[datetime],
    cancel_at_period_end: bool,
    issued_at: Optional[datetime] = None,
) -> str:
    """Mints a signed entitlement token. Raises EntitlementError (never a
    bare crypto exception) if the private key is missing or malformed."""
    if not private_key_b64:
        raise EntitlementError("PIPERSTITCH_LICENSE_PRIVATE_KEY is not configured — cannot sign an entitlement.")
    try:
        private_key = Ed25519PrivateKey.from_private_bytes(base64.b64decode(private_key_b64))
    except (binascii.Error, ValueError) as e:
        raise EntitlementError("PIPERSTITCH_LICENSE_PRIVATE_KEY is not a valid private key.") from e

    issued_at = issued_at or datetime.now(timezone.utc)
    payload = {
        "v": _VERSION,
        "cid": int(customer_id),
        "dev": device_id,
        "email": email,
        "status": status,
        "exp": _iso(expires_at),
        "iat": _iso(issued_at),
        "period_end": _iso(period_end) if period_end else None,
        "cancel_at_period_end": bool(cancel_at_period_end),
    }
    payload_bytes = json.dumps(payload, separators=(",", ":"), sort_keys=True).encode("utf-8")
    signature = private_key.sign(payload_bytes)
    token = f"{TOKEN_PREFIX}.{_b64url(payload_bytes)}.{_b64url(signature)}"

    # Self-check against the public key before the token leaves — a
    # mismatch here means the configured private key doesn't pair with
    # the public key baked into the shipped app, and every customer's
    # sign-in would silently fail. Far better to fail loudly here.
    parse_and_verify(token)
    return token


def parse_and_verify(token: str) -> Entitlement:
    """The same acceptance logic the app applies (EntitlementVerifier.swift):
    signature over the raw payload bytes first, then the fields."""
    parts = token.strip().split(".")
    if len(parts) != 3 or parts[0] != TOKEN_PREFIX:
        raise EntitlementError("That isn't a PiperStitch entitlement token.")
    try:
        payload_bytes = _unb64url(parts[1])
        signature = _unb64url(parts[2])
    except (binascii.Error, ValueError) as e:
        raise EntitlementError("That entitlement token is corrupted.") from e

    try:
        _public_key().verify(signature, payload_bytes)
    except InvalidSignature as e:
        raise EntitlementError("That entitlement token isn't valid.") from e

    try:
        payload = json.loads(payload_bytes)
        if payload.get("v") != _VERSION:
            raise EntitlementError("This entitlement was issued in a newer format than this service understands.")
        return Entitlement(
            customer_id=int(payload["cid"]),
            device_id=str(payload["dev"]),
            email=str(payload["email"]),
            status=str(payload["status"]),
            expires_at=_parse_iso(payload["exp"]),
            issued_at=_parse_iso(payload["iat"]),
            period_end=_parse_iso(payload["period_end"]) if payload.get("period_end") else None,
            cancel_at_period_end=bool(payload.get("cancel_at_period_end", False)),
            canonical=token.strip(),
        )
    except (KeyError, TypeError, ValueError, json.JSONDecodeError) as e:
        raise EntitlementError("That entitlement token is corrupted.") from e
